src/main/java/github/scarsz/discordsrv/api/SlashCommandManager.java

```diff
@@ -0,0 +1,142 @@
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * END
+ */
+
+package github.scarsz.discordsrv.api;
+
+import github.scarsz.discordsrv.DiscordSRV;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.exceptions.ErrorResponseException;
+import net.dv8tion.jda.api.exceptions.RateLimitedException;
+import net.dv8tion.jda.api.interactions.commands.build.CommandData;
+import net.dv8tion.jda.api.requests.ErrorResponse;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+public class SlashCommandManager {
+    private final Map<Plugin, PluginCommands> commandsMap = new HashMap<>();
+
+    /**
+     * Gets the Plugin Commands for a plugin
+     * @param plugin plugin to get commands of
+     * @return null if this plugin never added their commands here, or else the {@link PluginCommands}
+     */
+    public PluginCommands getCommands(Plugin plugin) {
+        return commandsMap.get(plugin);
+    }
+
+    public void reloadSlashCommands() {
+        Set<CommandData> commands = new HashSet<>();
+        for (PluginCommands value : commandsMap.values()) {
+            if (value.plugin.isEnabled()) {
+                commands.addAll(value.data);
+            }
+        }
+        if (!commands.isEmpty()) return; //no need to register if no commands will be registered
+        Map<Guild, RegistrationResult> errors = new HashMap<>();
+        for (Guild guild : DiscordSRV.getPlugin().getJda().getGuilds()) {
+            guild.updateCommands().addCommands(commands).queue(null, f -> errors.put(guild, RegistrationResult.getResult(f)));
+        }
+        printSlashRegistrationError(errors, commandsMap.keySet());
+    }
+
+    /**
+     * Sets the slash commands for a specific plugin, overrides old ones. this method does not register them, see {@link SlashCommandManager#reloadSlashCommands()}
+     * @param plugin Plugin that is registering the commands, whenever discordsrv is registering the commands it will not register the commands for that plugin if it is disabled
+     * @param commandData Commands to add
+     */
+    public void setCommands(@NotNull Plugin plugin, @NotNull CommandData... commandData) {
+        PluginCommands pluginCommands = commandsMap.get(plugin);
+        if (pluginCommands == null) {
+            pluginCommands = new PluginCommands(plugin, new HashSet<>(Arrays.asList(commandData)));
+            commandsMap.put(plugin, pluginCommands);
```

Comment:
Imo there's no need to reuse the PluginCommands object.

Make the `data` field final, and just replace the object in the map with the new set of commands. (i.e `commandsMap.put(plugin, new PluginCommands(plugin, commandData)`)

I think you should move the `HashSet` creation to the `PluginCommands` constructor